src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990QuantizationStateReader.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.enums.ScalarQuantizationType;
+import org.opensearch.knn.quantization.models.quantizationParams.ScalarQuantizationParams;
+import org.opensearch.knn.quantization.models.quantizationState.MultiBitScalarQuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.OneBitScalarQuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationStateReadConfig;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Reads quantization states
+ */
+public final class KNN990QuantizationStateReader {
+
+    /**
+     * Read quantization states and return list of fieldNames and bytes
+     * File format:
+     * Header
+     * QS1 state bytes
+     * QS2 state bytes
+     * Number of quantization states
+     * QS1 field number
+     * QS1 state bytes length
+     * QS1 position of state bytes
+     * QS2 field number
+     * QS2 state bytes length
+     * QS2 position of state bytes
+     * Position of index section (where QS1 field name is located)
+     * -1 (marker)
+     * Footer
+     *
+     * @param state the read state to read from
+     */
+    public static Map<String, byte[]> read(SegmentReadState state) throws IOException {
+        String quantizationStateFileName = getQuantizationStateFileName(state);
+        Map<String, byte[]> readQuantizationStateInfos = new HashMap<>();
+
+        try (IndexInput input = state.directory.openInput(quantizationStateFileName, IOContext.READ)) {
+            CodecUtil.retrieveChecksum(input);
+
+            int numFields = getNumFields(input);
+
+            // Read each field's metadata from the index section and then read bytes
+            for (int i = 0; i < numFields; i++) {
+                int fieldNumber = input.readInt();
+                int length = input.readInt();
+                long position = input.readVLong();
+                byte[] stateBytes = readStateBytes(input, position, length);
+                String fieldName = state.fieldInfos.fieldInfo(fieldNumber).getName();
+                readQuantizationStateInfos.put(fieldName, stateBytes);
+            }
+        } catch (Exception e) {
+            return readQuantizationStateInfos;
```

Comment:
I am assuming an exception is expected if the file is not found. In that case the catch all is risky here, change to catch a specific exception and then return `Collections.emptyMap()`. Creating a map on line 52 is not ideal as it allocates memory, `Collections.emptyMap` will not. Probably move the `new map()` to line 55. 

For all other exceptions log an error and please throw

Reply:
I am thinking to return empty map for all type of errors , and let's not load the cache , we really don't want to throw the excpetion and break and query flow. let it run usual.

Collections.emptyMap() also allocates memory in the same way new does , the only differnce is it's singleton and does not allocate every time , but here it does not matter.

Reply:
> I am thinking to return empty map for all type of errors , and let's not load the cache , we really don't want to throw the excpetion and break and query flow. let it run usual.

Why? Silent failures are not good

> Collections.emptyMap() also allocates memory in the same way new does , the only differnce is it's singleton and does not allocate every time , but here it does not matter.

For every request its creating new map which allocates memory, so we need that singleton

Reply:
Read functionality is getting called from NativeReader Constructor for both cases , like with quantization and without , without quantization we are returing empty map and with also in case of any error empty hash map basically don't load the hash map to load cache. Which in my case it's fine ..becuase flow is same. 

Segment reader is opened only once , and it's not going to open for each request , so it does not matter.

Reply:
1. This is a public function, so the code needs to be defensive
2. The segment reader is opened once per request. If there is enough requests we are unnecessarily allocating memory when not needed. So it might add up.

Lets make sure we aren't assigning memory where its not needed

Reply:
@Vikasht34 and @ryanbogan if we are catching the exception we should have logs too on what was the exception. Lets say if there is genuinely some error, then just eating up the exception is not correct.

More over, we are using IndexInput's exception to say if the file is present or not. Why not just get all the segmentfiles from the segment info and check if the desired file is present or not.

Reply:
I would suggest to have another check in both the read functions which check if QS file is present or not. and then safely return an empty map. and then we don't need to catch all the exceptions.

Reply:
Make Sense ...

Reply:
Will address in follow-up